File: algorithms/c/53.c
#include <stdio.h>

int maxSubArray(int* nums, int numsSize)
{
    if ( numsSize == 0 )
        return 0;
    int maxsum, thissum;
    maxsum = nums[0];
    thissum = 0;
    for ( int i = 0; i < numsSize; i++ )
    {
        thissum += nums[i];
        if ( thissum > maxsum )
            maxsum = thissum;
        if ( thissum < 0 )
            thissum = 0;
    }
    return maxsum;
}

int main()
{
    int arr[] = {-2,1,-3,4,-1,2,1,-5,4};
    printf("%d\n", maxSubArray(arr, sizeof(arr)/sizeof(arr[0])));
    return 0;
}

File: algorithms/python/53.py
class Solution(object):
    def maxSubArray(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        dp = nums[:]
        for i in range(1, len(dp)):
            dp[i] = max(nums[i], dp[i-1]+nums[i])
        return max(dp)
